### PR TITLE
Query: Remove ReLinq dependency from EntityQueryable

### PR DESCRIPTION
**Pull request by smitpatel**
Part of #12048

### CODE REVIEW

File: src/EFCore/Query/Internal/EntityQueryable`.cs [hunk 1]
```diff
@@ -5,49 +5,86 @@
...
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public Type ElementType => typeof(TResult);
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public Expression Expression { get; }
```
Comment (maumar) [inline src/EFCore/Query/Internal/EntityQueryable`.cs]: add virtual for those

File: src/EFCore/Query/Internal/EntityQueryable`.cs [hunk 2]
```diff
@@ -5,49 +5,86 @@
...
         ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
-        public EntityQueryable([NotNull] IAsyncQueryProvider provider, [NotNull] Expression expression)
-            : base(
-                Check.NotNull(provider, nameof(provider)),
-                Check.NotNull(expression, nameof(expression)))
+        public EntityQueryable([NotNull] IAsyncQueryProvider queryProvider, [NotNull] Expression expression)
         {
+            _queryProvider = queryProvider;
```
Comment (maumar) [inline src/EFCore/Query/Internal/EntityQueryable`.cs]: add null checks for args

File: src/EFCore/Query/Internal/QueryCompiler.cs [hunk 3]
```diff
@@ -249,29 +257,30 @@ public virtual Task<TResult> ExecuteAsync<TResult>(Expression query, Cancellatio
...
 
-            return database.CompileAsyncQuery<TResult>(queryModel);
+            var resultItemType
+                = (queryModel.GetOutputDataInfo()
+                      as StreamedSequenceInfo)?.ResultItemType
+                  ?? typeof(TResult).TryGetSequenceType();
+
+            try
+            {
+                return (Func<QueryContext, TResult>)CompileAsyncQueryMethod
```
Comment (maumar) [inline src/EFCore/Query/Internal/QueryCompiler.cs]: eww, reflection

File: src/EFCore/Query/Internal/EntityQueryable`.cs (hunk 1 above)
Comment (smitpatel) [reply src/EFCore/Query/Internal/EntityQueryable`.cs]: non-virtual. What is the use-case you think, we should allow this to be overridable? Further it's getter-only, ctor set.

File: src/EFCore/Query/Internal/QueryCompiler.cs (hunk 3 above)
Comment (smitpatel) [reply src/EFCore/Query/Internal/QueryCompiler.cs]: One should not complain about reflection on async code path :trollface:

Comment (smitpatel) [reply src/EFCore/Query/Internal/QueryCompiler.cs]: Sync path also uses reflection.

**Pull request merged**